blog.md

```diff
@@ -0,0 +1,215 @@
+will communicate using a UNIX Domain Socket (UDS) with each plugin and provide
+them with all required event data. For example, events can be container or pod
+sandbox _creation_, _stopping_ or _removal_, while corresponding data are the
+_name_, _namespace_ or corresponding _annotations_.
+
+[api]: https://github.com/containerd/nri/blob/eaf78a9/pkg/api/api.proto
+[ttrpc]: https://github.com/containerd/ttrpc
+[grpc]: https://grpc.io
+
+On one hand plugins written as daemons have the benefit of persisting the
+current state out of the box, while on the other hand they come with a
+performance and management overhead. For that reason, the NRI also supports
+[OCI hook][hook]-like [binary plugins][v010] which get executed for each event.
+Combining the concept of small binary plugins with a universal standard like
+[WebAssembly (Wasm)][wasm] empowers the NRI to run on the edge and universally
+on all imaginable platforms.
+
+[hook]: https://github.com/opencontainers/runtime-spec/blob/9ceba9f/config.md#posix-platform-hooks
+[v010]: https://github.com/containerd/nri/tree/693d64e/plugins/v010-adapter
+[wasm]: https://webassembly.org
+
+## How it works
+
+The required change for the NRI landed with [Pull Request
+containerd/nri#121][pr121]. This change adds a [go-plugin][go-plugin] mechanism
+to the NRI. Each plugin gets compiled to Wasm, which means that it is
+size-efficient, memory-safe, automatically sandboxed and highly portable out of
+the box! The plugin system works in the same way as the NRI by using [Protocol
+Buffers][protobuf]. This means, that the NRI can reuse the existing API for
+ttRPC, while the communication will happen in memory and not over the Remote
+Procedure Call (RPC).
+
+[pr121]: https://github.com/containerd/nri/pull/121
+[go-plugin]: https://github.com/knqyf263/go-plugin
+[protobuf]: https://protobuf.dev
+[tinygo]: https://tinygo.org
+
+One key benefit is that WebAssembly is designed as a portable compilation target
+for programming languages. Plugins compiled to Wasm can be used anywhere, which
+means that there is no requirement for multi architecture binaries. Beside that,
+the Wasm stack machine is designed to be encoded in a size and time efficient
+binary format, which make them great targets for binary execution.
+
+## Demo
+
+Unfortunately, the native golang (`go`) compiler does not have full WebAssembly
+support yet, which means the plugins have to be compiled using the alternative
+[tinygo][tinygo] compiler. An [example Wasm plugin][example] within the NRI
+repository, can be compiled locally using:
+
+```bash
+make $(pwd)/build/bin/wasm
+```
+
+Or within a container image:
+
+```bash
+make $(pwd)/build/bin/wasm TINYGO_DOCKER=1
+```
+
+In the future it may be possible to cross compile plugins using `GOOS=wasip1
+GOARCH=wasm go build`, but that is not implemented yet (see
+[knqyf263/go-plugin#58][go-plugin-58]).
+
+[example]: https://github.com/containerd/nri/blob/dd57194/plugins/wasm/plugin.go
```

Comment:
Github gives a warning about this link:

This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.

...it'll be probably safer to link to a commit in the main branch, like you have in the other links below.

Reply:
I'll update the link, thank you!